### PR TITLE
Feat: Add Resource Locking to ALZ Bicep Modules

### PR DESCRIPTION
**Pull request by DaFitRobsta**
# Overview/Summary

Adds the support for resource locking in all modules for resources that support resource locks. 
See [AB#30524](https://dev.azure.com/CSUSolEng/2a9d5a5a-8998-4ad0-81c8-ef3045e4da97/_workitems/edit/30524) for additional details of feature request.
In response to #608 

## This PR fixes/adds/changes/removes

- Allows for setting resource locking globally or at each individual resource within the module.
- Resource locking at the subscription level is not supported in this release.

### Breaking Changes

None

## Testing Evidence

Each module that was updated to support Resource Locking was tested with successful deployments. Testing included setting the global parameter to different values as individual resources to validate resource locking logic.

## As part of this Pull Request I have

- [x] Read the [Contribution Guide](https://github.com/Azure/ALZ-Bicep/wiki/Contributing) and ensured this PR is compliant with the guide
- [x] Ensured the resource API versions in `.bicep` file/s I am adding/editing are using the latest API version possible
- [x] Checked for duplicate [Pull Requests](https://github.com/Azure/ALZ-Bicep/pulls)
- [x] Associated it with relevant [GitHub Issues](https://github.com/Azure/ALZ-Bicep/issues)
- [x] *(ALZ Bicep Core Team Only)* Associated it with relevant [ADO Items](https://aka.ms/alz/bicep/backlog)
- [x] Ensured my code/branch is up-to-date with the latest changes in the `main` [branch](https://github.com/Azure/ALZ-Bicep/tree/main)
- [x] Performed testing and provided evidence.
- [ ] Updated one or more of the following tests *(if required)*
  - [Unit](https://github.com/Azure/ALZ-Bicep/blob/main/.github/workflows/bicep-build-to-validate.yml)
  - [Linting](https://github.com/Azure/ALZ-Bicep/tree/main/.github/workflows)
  - [E2E (End-To-End)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/bicep-build-to-validate.yml)
  - [ValidateAzCloud (Base validation in Azure Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/base-unit-validate.yml)
  - [ValidateMcCloud (Base validation in Azure China Cloud)](https://github.com/Azure/ALZ-Bicep/blob/main/tests/pipelines/mc-base-unit-validate.yml)
- [x] Updated relevant and associated documentation (e.g. Contribution Guide, Module READMEs, Wiki Docs etc.)
- [ ] If relevant, created or updated Code Tours [here](https://github.com/Azure/ALZ-Bicep/blob/main/.vscode/tours)

### CODE REVIEW

**Comment by oZakari** [comment]
Thanks @DaFitRobsta! Will go through this PR later this week. 👍🏼

**Review by oZakari — CHANGES_REQUESTED** [review]
Hey @DaFitRobsta, for the resource group module I see there is a new module called resourceGroupLock.bicep. Can we instead just use the resource lock provider as you have done for the other modules instead?

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep [hunk 1]
```diff
@@ -175,6 +192,15 @@ resource resPrivateDnsZones 'Microsoft.Network/privateDnsZones@2020-06-01' = [fo
   tags: parTags
 }]
 
+resource lock 'Microsoft.Authorization/locks@2020-05-01' = [for (privateDnsZone, index) in varPrivateDnsZonesMerge: if (parResourceLockConfig.kind != 'None') {
```
Comment (oZakari) [inline infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:195]:
```suggestion
resource resPrivateDnsZonesLock 'Microsoft.Authorization/locks@2020-05-01' = [for (privateDnsZone, index) in varPrivateDnsZonesMerge: if (parResourceLockConfig.kind != 'None') {
```

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep [hunk 2]
```diff
@@ -188,6 +214,15 @@ resource resVirtualNetworkLink 'Microsoft.Network/privateDnsZones/virtualNetwork
   tags: parTags
 }]
 
+resource lockLink 'Microsoft.Authorization/locks@2020-05-01' = [for (privateDnsZone, index) in varPrivateDnsZonesMerge: if (!empty(parVirtualNetworkIdToLink) && !empty(parResourceLockConfig ?? {}) && parResourceLockConfig.kind != 'None') {
```
Comment (oZakari) [inline infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:217]:
```suggestion
resource resVirtualNetworkLinkLock 'Microsoft.Authorization/locks@2020-05-01' = [for (privateDnsZone, index) in varPrivateDnsZonesMerge: if (!empty(parVirtualNetworkIdToLink) && !empty(parResourceLockConfig ?? {}) && parResourceLockConfig.kind != 'None') {
```

File: infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep [hunk 3]
```diff
@@ -201,6 +236,15 @@ resource resVirtualNetworkLinkFailover 'Microsoft.Network/privateDnsZones/virtua
   tags: parTags
 }]
 
+resource lockFailoverLink 'Microsoft.Authorization/locks@2020-05-01' = [for (privateDnsZone, index) in varPrivateDnsZonesMerge: if (!empty(parVirtualNetworkIdToLinkFailover) && !empty(parResourceLockConfig ?? {}) && parResourceLockConfig.kind != 'None') {
```
Comment (oZakari) [inline infra-as-code/bicep/modules/privateDnsZones/privateDnsZones.bicep:239]:
```suggestion
resource resVirtualNetworkLinkFailoverLock 'Microsoft.Authorization/locks@2020-05-01' = [for (privateDnsZone, index) in varPrivateDnsZonesMerge: if (!empty(parVirtualNetworkIdToLinkFailover) && !empty(parResourceLockConfig ?? {}) && parResourceLockConfig.kind != 'None') {
```

File: infra-as-code/bicep/modules/publicIp/publicIp.bicep [hunk 4]
```diff
@@ -39,6 +56,15 @@ resource resPublicIp 'Microsoft.Network/publicIPAddresses@2023-02-01' ={
   properties: parPublicIpProperties
 }
 
+resource lock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(parResourceLockConfig ?? {}) && parResourceLockConfig.kind != 'None') {
```
Comment (oZakari) [inline infra-as-code/bicep/modules/publicIp/publicIp.bicep:59]:
```suggestion
resource resPublicIpLock 'Microsoft.Authorization/locks@2020-05-01' = if (!empty(parResourceLockConfig ?? {}) && parResourceLockConfig.kind != 'None') {
```

File: infra-as-code/bicep/modules/vwanConnectivity/parameters/vwanConnectivity.parameters.all.json [hunk 5]
```diff
@@ -3,7 +3,7 @@
   "contentVersion": "1.0.0.0",
   "parameters": {
     "parLocation": {
-      "value": "eastus"
+      "value": "westus3"
```
Comment (oZakari) [inline infra-as-code/bicep/modules/vwanConnectivity/parameters/vwanConnectivity.parameters.all.json:6]: Please change this back to eastus.

**Review by oZakari — CHANGES_REQUESTED** [review]

File: infra-as-code/bicep/modules/spokeNetworking/parameters/spokeNetworking.parameters.all.json [hunk 6]
```diff
@@ -21,7 +21,7 @@
       "value": []
     },
     "parNextHopIpAddress": {
-      "value": ""
+      "value": "192.168.2.4"
```
Comment (oZakari) [inline infra-as-code/bicep/modules/spokeNetworking/parameters/spokeNetworking.parameters.all.json:24]: Please revert.

**Review by oZakari — CHANGES_REQUESTED** [review]

**Comment by DaFitRobsta** [comment]
> Hey @DaFitRobsta, for the resource group module I see there is a new module called resourceGroupLock.bicep. Can we instead just use the resource lock provider as you have done for the other modules instead?

Hey @oZakari,
Because the resource group module has a targetScope set to the subscription level, if placing the resource lock within the module, it locks the subscription and not the resource group. I used this site, https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/lock-resources?tabs=bicep#template, as an example to work around the issue. It's my understanding that I have to call a sub-module in order to set the resource lock at the resource group level. If you have other methods to execute this without the extra module, please let me know.

File: infra-as-code/bicep/modules/spokeNetworking/parameters/spokeNetworking.parameters.all.json (hunk 6 above)
Comment (DaFitRobsta) [reply infra-as-code/bicep/modules/spokeNetworking/parameters/spokeNetworking.parameters.all.json:24]: You got it

**Comment by oZakari** [comment]
> > Hey @DaFitRobsta, for the resource group module I see there is a new module called resourceGroupLock.bicep. Can we instead just use the resource lock provider as you have done for the other modules instead?
> 
> Hey @oZakari, Because the resource group module has a targetScope set to the subscription level, if placing the resource lock within the module, it locks the subscription and not the resource group. I used this site, https://learn.microsoft.com/en-us/azure/azure-resource-manager/management/lock-resources?tabs=bicep#template, as an example to work around the issue. It's my understanding that I have to call a sub-module in order to set the resource lock at the resource group level. If you have other methods to execute this without the extra module, please let me know.

Ah, that makes perfect sense. Thanks for the clarification on that!

**Comment by oZakari** [comment]
Hey @DaFitRobsta, the ValidateAzCloud pipeline failed due to parPublicIPLock not being declared within the hub networking module. Could you add that in when you get chance. After adding the fix, you can run put /azp run validateazcloud in a comment to re-trigger the pipeline.

**Review by oZakari — APPROVED** [review]

**Pull request merged**